### ISSUE TITLE
Feature Request: Custom module settings in Module.txt

### ISSUE THREAD

**Issue by dcinzona** [OrchardCMS/Orchard#5871]
Instead of storing custom module settings in the root web.config, it would be nice to support custom settings in module.txt.  This would allow for updating settings without restarting the web application.  Or maybe, instead, add an module.config file that allows for transforms for deploying to different environments.  Then we would create a new interface that could read from the module.config much like App Settings.

The reason I bring this up is that I've seen first hand and through questions in Gitter, where devs want to store custom settings within a module, and while this is possible, retrieving those settings requires some boiler plate code.  It would be nice if this was built in as a feature.

One example could be if a module needs to talk to a different database, the DB connection string could be in module.config and retrieved via `IModuleConfig.GetSetting("myConnString")`.  You could also extend this to allow for pulling settings from other modules: `IModuleConfig["AnotherModule"].GetSetting("someSetting")`.

Thoughts?

**Comment by Piedone** [comment]
I fail to see the point: modules are self-contained packages that shouldn't be modified (just as Orchard's own code shouldn't be modified) but just extended. Now with the root Web.config we don't have much choice, it should be included in Orchard and you should be able to modify it (since this is how ASP.NET works), but adding another layer of modding with module Web.configs doesn't seem the right thing to do for me. This is basically the same as storing settings in constants in the module's C# code and changing those.

**Comment by dcinzona** [comment]
True, but changing settings in those scenarios require app restarts or deployments.  In this scenario, there's the ability to specify the config for custom module.  So in our situation, we could change the API url if we needed to, without forcing an app restart.

**Comment by dcinzona** [comment]
The only other way to do it without requiring app restarts is by creating settings in the admin UI, which is totally viable, but cumbersome

**Comment by sebastienros** [comment]
To prevent the custom configurations from being altered during a deployment, we need a separate file with all the custom settings, and configuration service to read this file per tenant.

@dcinzona is also suggesting to have a sample file in the modules folders so that an update to the file with new settings wouldn't break existing sites, and would still list all available settings and sample values.
